### ISSUE TITLE
Unable to add Things via text config after they previously removed

### ISSUE THREAD

**Issue by QNimbus** [openhab/openhab-core#1042]
I noticed a problem in OpenHAB2.5.0M1 & OpenHAB2.5.0M2.

When commenting out a 'Thing' via a text file and later re-adding that same 'Thing' by removing the comments - the text file gets parsed correctly but the 'Thing' is never created (again).

Just to clarify: This is not an issue with a specific binding (in this case the Astro binding). I just used this for testing purposes because it's a small, stable binding than be installed and configured quickly. This issue also applies to other bindings and their config. 

To reproduce I installed the 'Astro' binding and created a `astro.things ` text file containing the following configuration:

```
Thing astro:sun:home        "VT_Astro_Sun"          @   "Virtual"       [ geolocation="6.524379,3.379206", interval=60 ] 
{
    Channels:
        Type rangeEvent : rise#event [
            offset=0
        ]
        Type rangeEvent : set#event [
            offset=0
        ]        
}
Thing astro:moon:home       "VT_Astro_Moon"         @   "Virtual"       [ geolocation="6.524379,3.379206", interval=60 ]
```

This works as expected and creates two Things in OpenHAB.

Next I comment out the 'VT_Astro_Moon' in the 'Thing' config and it's handler gets disposed and the Thing disappears from OpenHAB as expected.

When I remove the previously added comments to have OpenHAB (re)create the 'VT_Astro_Moon' thing nothing gets created.

I'm not sure if the problem lies with the original Thing not properly getting disposed or with the instantiation of the Thing/ThingHandler when I try to add it again.

I tested this on an OpenHAB build running on Windows 10 as well as within a Docker container on RancherOS. All builds were freshly installed and contained no other bindings or configurations other than what is there by default.

~~I will try to see if I can reproduce this on 2.4.0 and the 2.5.0-SNAPSHOT as well~~

EDIT: I was able to reproduce this behavior in 2.4.0 as well as the latest SNAPSHOT. The only ways around this issue are: 
- To rename the 'thing_id' part of the configuration. (`Thing <binding_id>:<type_id>:<thing_id>`)
  - OR-
- To restart the entire OpenHAB instance

**Comment by lolodomo** [comment]
After the first step (removing), can you please run the following command in the console `smarthome:things list` and check if the removed thing is still present and if it is what is its status ?
Please also run the command `smarthome:links list` to check if some channel links for this thing are still there ?

**Comment by QNimbus** [comment]
@lolodomo 

After I installed and configured the 'Astro' binding the output of `smarthome:things list` is : 

```
openhab> smarthome:things list
astro:sun:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Sun, Bridge=null)
astro:moon:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Moon, Bridge=null)
```

After removing the 'moon' Thing the output is : 

```
openhab> smarthome:things list
astro:sun:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Sun, Bridge=null)
```
 The command `smarthome:links list` returns and empty result since I have no items (or channels) linked.

When I remove the comments from the .things config and re-add the 'moon' item the result is :  

```
openhab> smarthome:things list
astro:sun:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Sun, Bridge=null)
```

as opposed to the expected result : 

```
openhab> smarthome:things list
astro:sun:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Sun, Bridge=null)
astro:moon:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Moon, Bridge=null)
```

**Comment by lolodomo** [comment]
Ok so the problem is not linked channels and apparently the thing was really removed from the registry. 

When uncommenting, do you see a log that tells you the file is refreshed or reloaded? 

Is your think file local or on a remote share?

I will try to reproduce.

**Comment by QNimbus** [comment]
> When uncommenting, do you see a log that tells you the file is refreshed or reloaded?

Yes I do. After commenting out the 'Moon' thing the log file prints out ==>

```
15:27:28.423 [INFO ] [del.core.internal.ModelRepositoryImpl] - Refreshing model 'astro.things'
15:27:28.437 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from ONLINE to UNINITIALIZED
15:27:28.473 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
```

>Is your think file local or on a remote share?

The `astro.things` file I used to configure the Astro things is a **local** file in the **conf** folder.

**Comment by lolodomo** [comment]
Could you please enable these logs ? `log:set TRACE org.eclipse.smarthome.model.thing`
I am interested by the logs when the thing file is refreshed after you uncommented the thing.

**Comment by QNimbus** [comment]
After installing the Astro binding on a fresh OpenHAB installation : 

```
20:02:53.159 [INFO ] [del.core.internal.ModelRepositoryImpl] - Loading model 'astro.things'
20:02:53.164 [DEBUG] [l.thing.internal.GenericThingProvider] - Read things from model 'astro.things'
20:02:53.198 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:sun' with UID 'astro:sun:home.
20:02:53.255 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:sun
20:02:53.326 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:moon' with UID 'astro:moon:home.
20:02:53.347 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:moon
20:02:53.400 [DEBUG] [l.thing.internal.GenericThingProvider] - Adding thing 'astro:sun:home' from model 'astro.things'.
20:02:53.454 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:sun:home' changed from UNINITIALIZED to INITIALIZING
20:02:53.594 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:sun:home' changed from INITIALIZING to ONLINE
20:02:53.637 [INFO ] [penhab.binding.astro.internal.job.Job] - Scheduled Astro event-jobs for thing astro:sun:home
20:02:53.670 [DEBUG] [l.thing.internal.GenericThingProvider] - Adding thing 'astro:moon:home' from model 'astro.things'.
20:02:53.704 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from UNINITIALIZED to INITIALIZING
20:02:53.729 [INFO ] [penhab.binding.astro.internal.job.Job] - Scheduled Astro event-jobs for thing astro:moon:home
20:02:53.738 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from INITIALIZING to ONLINE
20:02:54.777 [INFO ] [del.core.internal.ModelRepositoryImpl] - Refreshing model 'astro.things'
20:02:54.789 [DEBUG] [l.thing.internal.GenericThingProvider] - Read things from model 'astro.things'
20:02:54.811 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:sun' with UID 'astro:sun:home.
20:02:54.814 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:sun
20:02:54.838 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:moon' with UID 'astro:moon:home.
20:02:54.855 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:moon
```

Right after commenting out the 'Moon' thing : 

```
20:05:03.938 [INFO ] [del.core.internal.ModelRepositoryImpl] - Refreshing model 'astro.things'
20:05:03.946 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from ONLINE to UNINITIALIZED
20:05:03.984 [INFO ] [ome.event.ThingStatusInfoChangedEvent] - 'astro:moon:home' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
20:05:04.006 [DEBUG] [l.thing.internal.GenericThingProvider] - Read things from model 'astro.things'
20:05:04.014 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:sun' with UID 'astro:sun:home.
20:05:04.016 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:sun
```

So far so good..... 

Now I delete the comments in front of the 'Moon' thing : 

```
20:06:08.009 [INFO ] [del.core.internal.ModelRepositoryImpl] - Refreshing model 'astro.things'
20:06:08.009 [INFO ] [del.core.internal.ModelRepositoryImpl] - Refreshing model 'astro.things'
20:06:08.017 [DEBUG] [l.thing.internal.GenericThingProvider] - Read things from model 'astro.things'
20:06:08.017 [DEBUG] [l.thing.internal.GenericThingProvider] - Read things from model 'astro.things'
20:06:08.020 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:sun' with UID 'astro:sun:home.
20:06:08.020 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:sun' with UID 'astro:sun:home.
20:06:08.024 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:sun
20:06:08.024 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:sun
20:06:08.056 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:moon' with UID 'astro:moon:home.
20:06:08.056 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing for type 'astro:moon' with UID 'astro:moon:home.
20:06:08.058 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:moon
20:06:08.058 [TRACE] [l.thing.internal.GenericThingProvider] - Creating thing from specific ThingHandlerFactory org.openhab.binding.astro.internal.AstroHandlerFactory@a98a7b4 for thingType astro:moon
```

And to verify I ran the `smarthome:things list` command again : 

```
openhab> smarthome:things list
astro:sun:home (Type=Thing, Status=ONLINE, Label=VT_Astro_Sun, Bridge=null)
```

This was all done on a fresh installation of 2.5.0M3 with nothing configured or installed except for the `Astro` binding (and it's `.things` config file). Is this something you can reproduce as well? I know that a buddy of mine was able to reproduce this same issue at my request on hist OpenHAB installation. I've asked him to chime in here as well.

If there is anything else I can do to provide more insight in this issue, please let me know!

**Comment by lolodomo** [comment]
Strange that all logs are in double. 
They are not in double when you comment the thing but are in double when you uncomment. 
Can you confirm this is always the case?

**Comment by ByKaj** [comment]
I have the same behavior on openHAB 2.5.0.M1. If there's some info I can provide, please let me know. For now I have the same results as @QNimbus.

**Comment by cweitkamp** [comment]
I can confirm your issue and I am able to reproduce it.

@lolodomo Do you yet have an idea how to fix it? It looks like the `GenericThingProvider` uses an internal `thingsMap` to track all things from a specific file. New things will be added and all things will be removed if the file will not be present anymore BUT if only a single thing (or a part of things) will be removed they remain in the internal map (see https://github.com/openhab/openhab-core/blob/c9232d9651638a48accf0e259f3d4e63421026f1/bundles/org.openhab.core.model.thing/src/org/eclipse/smarthome/model/thing/internal/GenericThingProvider.xtend#L486 ff).

**Comment by QNimbus** [comment]
> Strange that all logs are in double.
> They are not in double when you comment the thing but are in double when you uncomment.
> Can you confirm this is always the case?

No that seems to have been a one-time glitch - I just tried the same procedure again and just got single log entries. That doesn't seem to be part of the issue.

**Comment by lolodomo** [comment]
@cweitkamp : My immediate intuition was a possible problem with this thingMap too. 
I am not very familiar with the Xtend code. If you want to take this one, I let you do.

**Comment by QNimbus** [comment]
@cweitkamp 
> Do you yet have an idea how to fix it? It looks like the `GenericThingProvider` uses an internal `thingsMap` to track all things from a specific file. New things will be added and all things will be removed if the file will not be present anymore BUT if only a single thing (or a part of things) will be removed they remain in the internal map.

Is the same mechanism used for `.items` files, because they seem to work fine in this respect?

**Comment by cweitkamp** [comment]
> Is the same mechanism used for .items files, because they seem to work fine in this respect?

Yes, the logic should be similar. But the implementation is different,

I submitted a fix for this issue (see #1058).

**Comment by QNimbus** [comment]
@cweitkamp @lolodomo 

I appreciate the way you both jumped on this and created a fix for this - thanks! This is what make makes the OpenHAB community great!
